fedimint-client/src/db.rs

```diff
@@ -262,3 +276,173 @@ impl_db_lookup!(
+            )));
+        }
+
+        while current_db_version < target_db_version {
+            let new_states = if let Some(migration) = migrations.get(&current_db_version) {
+                info!(target: LOG_DB, "Migrating module {kind} from {current_db_version} to {target_db_version}");
+                let active_states = get_active_states(&mut global_dbtx.to_ref_nc()).await;
+                let inactive_states = get_inactive_states(&mut global_dbtx.to_ref_nc()).await;
+
+                migration(
+                    &mut global_dbtx
+                        .to_ref_with_prefix_module_id(module_instance_id)
+                        .into_nc(),
+                    module_instance_id,
+                    active_states,
+                    inactive_states,
+                )
+                .await?
+            } else {
+                warn!("Missing client db migration for version {current_db_version}");
+                None
+            };
+
+            // If the client migration returned new states, a state machine migration has
+            // occurred, and the new states need to be persisted to the database.
+            if let Some((new_active_states, new_inactive_states)) = new_states {
+                persist_active_states(&mut global_dbtx.to_ref_nc(), new_active_states).await;
+                persist_inactive_states(&mut global_dbtx.to_ref_nc(), new_inactive_states).await;
+            }
+
+            current_db_version.increment();
+            global_dbtx
+                .insert_entry(&DatabaseVersionKey(module_instance_id), &current_db_version)
+                .await;
+        }
+
+        current_db_version
+    } else {
+        target_db_version
+    };
+
+    global_dbtx.commit_tx_result().await?;
+    info!(target: LOG_DB, "{} module db version: {}", kind, db_version);
+    Ok(())
+}
+
+/// Reads all active states from the database and returns `Vec<DynState>`.
+pub async fn get_active_states(dbtx: &mut DatabaseTransaction<'_>) -> Vec<DynState> {
+    dbtx.find_by_prefix(&ActiveStateKeyPrefix::new())
+        .await
+        .map(|(state, _)| state.state)
+        .collect::<Vec<_>>()
+        .await
+}
+
+/// Reads all inactive states from the database and returns `Vec<DynState>`.
+pub async fn get_inactive_states(dbtx: &mut DatabaseTransaction<'_>) -> Vec<DynState> {
+    dbtx.find_by_prefix(&InactiveStateKeyPrefix::new())
+        .await
+        .map(|(state, _)| state.state)
+        .collect::<Vec<_>>()
+        .await
+}
+
+/// Persists new active states by first removing all current active states, and
+/// re-writing with the new set of active states. `new_active_states` is
+/// expected to contain all active states, not just the newly created states.
+pub async fn persist_active_states(
+    dbtx: &mut DatabaseTransaction<'_>,
+    new_active_states: Vec<DynState>,
+) {
+    // Remove all existing active states
+    dbtx.remove_by_prefix(&ActiveStateKeyPrefix::new()).await;
```

Comment:
With filtering by module_instance_id we should be able to delete only module's own states.

Reply:
Made an issue, assigned myself https://github.com/fedimint/fedimint/issues/4287